nifi-extension-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/stream/record/StateHandlerStrategy.java

```diff
@@ -0,0 +1,111 @@
+import org.apache.nifi.schema.access.SchemaNotFoundException;
+import org.apache.nifi.serialization.record.Record;
+import org.apache.nifi.serialization.record.RecordSchema;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+
+/**
+ * StateHandlerStrategy is responsible for managing the state of FlowFiles created for Records processed in a single batch. In general this class decides what should happen when a new RecordSchema
+ * is encountered — whether previous State should be completed and new state created or only a new state should be created.
+ */
+class StateHandlerStrategy {
+
+    private final SchemaDifferenceHandlingStrategy strategy;
+    private final StateInitializerAction stateInitializerAction;
+    private final StateFinalizerAction stateFinalizerAction;
+    private final Map<RecordSchema, FlowFileState> activeStateMap = new HashMap<>();
+
+    StateHandlerStrategy(final SchemaDifferenceHandlingStrategy strategy, final StateInitializerAction stateInitializerAction, final StateFinalizerAction stateFinalizerAction) {
+        this.strategy = strategy;
+        this.stateInitializerAction = stateInitializerAction;
+        this.stateFinalizerAction = stateFinalizerAction;
+    }
+
+    FlowFileState getOrCreate(final Record record, final BatchProcessingContext flowFileContext) throws FlowFileCompletionException, IOException, SchemaNotFoundException {
+        return switch (strategy) {
+            case ROLL_FLOW_FILES -> getOrFinalizeAndCreateNewState(record, flowFileContext);
+            case GROUP_FLOW_FILES -> getOrCreateNewState(record, flowFileContext);
+        };
+    }
+
+    private FlowFileState getOrFinalizeAndCreateNewState(final Record record, final BatchProcessingContext flowFileContext) throws FlowFileCompletionException, IOException, SchemaNotFoundException {
+        final FlowFileState previousState = activeStateMap.get(record.getSchema());
+        if (previousState != null) {
+            return previousState;
+        }
+        final FlowFileState previousStateForDifferentSchema = pop();
+        if (previousStateForDifferentSchema != null) {
+            stateFinalizerAction.complete(previousStateForDifferentSchema, flowFileContext);
+        }
+        final FlowFileState newState = stateInitializerAction.init(record, flowFileContext);
+        activeStateMap.put(record.getSchema(), newState);
+        return newState;
+    }
+
+    private FlowFileState getOrCreateNewState(final Record record, final BatchProcessingContext flowFileContext) throws IOException, SchemaNotFoundException {
+        final FlowFileState previousState = activeStateMap.get(record.getSchema());
+        if (previousState != null) {
+            return previousState;
+        }
+        final FlowFileState newState = stateInitializerAction.init(record, flowFileContext);
+        activeStateMap.put(record.getSchema(), newState);
+        return newState;
+    }
```

Comment:
The two different methods appear to be the same, except for the `stateFinalizerAction` call, I recommend collapsing them and just checking for the `Roll FlowFiles` strategy, unless there is some other key difference I am missing.